packages/assets-controllers/src/MultichainTokensRatesController/MultichainTokensRatesController.ts

```diff
@@ -0,0 +1,391 @@
+
+const SWIFT_ISO4217_USD = 'swift:0/iso4217:USD';
+
+/**
+ * State used by the MultiChainTokensRatesController to cache token conversion rates.
+ */
+export type MultichainTokensRatesControllerState = ConversionRatesWrapper;
+
+/**
+ * Returns the state of the MultiChainTokensRatesController.
+ */
+export type MultichainTokensRatesControllerGetStateAction =
+  ControllerGetStateAction<
+    typeof controllerName,
+    MultichainTokensRatesControllerState
+  >;
+
+/**
+ * Action to update the rates of all supported tokens.
+ */
+export type MultichainTokensRatesControllerUpdateRatesAction = {
+  type: `${typeof controllerName}:updateTokensRates`;
+  handler: MultiChainTokensRatesController['updateTokensRates'];
+};
+
+/**
+ * Event emitted when the state of the MultiChainTokensRatesController changes.
+ */
+export type MultichainTokensRatesControllerStateChange =
+  ControllerStateChangeEvent<
+    typeof controllerName,
+    MultichainTokensRatesControllerState
+  >;
+
+/**
+ * Actions exposed by the MultiChainTokensRatesController.
+ */
+export type MultichainTokensRatesControllerActions =
+  | MultichainTokensRatesControllerGetStateAction
+  | MultichainTokensRatesControllerUpdateRatesAction;
+
+/**
+ * Events emitted by MultiChainTokensRatesController.
+ */
+export type MultichainTokensRatesControllerEvents =
+  MultichainTokensRatesControllerStateChange;
+
+/**
+ * Actions that this controller is allowed to call.
+ */
+export type AllowedActions =
+  | GetAllSnaps
+  | HandleSnapRequest
+  | AccountsControllerListMultichainAccountsAction
+  | AccountsControllerGetStateAction;
+
+/**
+ * Events that this controller is allowed to subscribe to.
+ */
+export type AllowedEvents =
+  | AccountsControllerAccountRemovedEvent
+  | KeyringControllerLockEvent
+  | KeyringControllerUnlockEvent
+  | AccountsControllerSelectedAccountChangeEvent;
+
+/**
+ * Messenger type for the MultiChainTokensRatesController.
+ */
+export type MultichainTokensRatesControllerMessenger =
+  RestrictedControllerMessenger<
+    typeof controllerName,
+    MultichainTokensRatesControllerActions | AllowedActions,
+    MultichainTokensRatesControllerEvents | AllowedEvents,
+    AllowedActions['type'],
+    AllowedEvents['type']
+  >;
+
+/**
+ * The input for starting polling in MultiChainTokensRatesController.
+ */
+export type MultiChainRatesPollingInput = {
+  accountId: string;
+};
+
+const metadata = {
+  conversionRates: { persist: true, anonymous: true },
+};
+
+/**
+ * Controller that manages multichain token conversion rates.
+ *
+ * This controller polls for token conversion rates and updates its state.
+ */
+export class MultiChainTokensRatesController extends StaticIntervalPollingController<MultiChainRatesPollingInput>()<
+  typeof controllerName,
+  MultichainTokensRatesControllerState,
+  MultichainTokensRatesControllerMessenger
+> {
+  readonly #mutex = new Mutex();
+
+  #accountId: AccountsControllerState['internalAccounts']['selectedAccount'];
```

Comment:
Should we still be relying on the currently selected account ?

Reply:
By polling only the currently selected account, we avoid sending requests for non-selected accounts and prevent overloading the price API with data that isn’t visible in the UI. 
this is why we rely on selected account only

Reply:
That will not work when we get rid of the selected account concept.

I know it's not the case right now and a lot of things still relies on this in the extension but we already started not relying on it in multiple places.

This is going to cause problems when we're trying to display data for an account that's not currently selected.

Reply:
alright, I understand your point. I'll make the necessary changes to remove reliance on the selected account, implement the modifications, and then get back to you for a review.

Reply:
Potentially a dumb question, correct me if I am wrong, but if we only poll rates for the selected account will the fiat value not be out of date for other accounts in the account selector for instance?

Reply:
yes if we choose to poll only on selected account , the value for non selected account will be out of date indeed , this is why we trigger updateRates when we switch ( let me know if this answer your question ).
I'm going to change the logic now , the polling will be done across all added accounts